File: core/src/main/java/com/hgianastasio/biblioulbrav2/core/base/boundaries/Mapper.java
package com.hgianastasio.biblioulbrav2.core.base.boundaries;

import java.util.LinkedList;
import java.util.List;

/**
 * Created by heitorgianastasio on 4/24/17.
 */

public abstract class Mapper<I,O> {

    public abstract O transform(I input);


    public final List<O> trasform(List<I> input){
        List<O> result = new LinkedList<>();
        for (I inputObj : input)
            result.add(transform(inputObj));
        return result;
    }


    public abstract I transformBack(O input);


    public final List<I> tansformBack(List<O> input){
        List<I> result = new LinkedList<>();
        for (O inputObj : input)
            result.add(transformBack(inputObj));
        return result;

    }
}

File: core/src/main/java/com/hgianastasio/biblioulbrav2/core/base/boundaries/Mapper.kt
package com.hgianastasio.biblioulbrav2.core.base.boundaries

import java.util.*

/**
 * Created by heitorgianastasio on 4/24/17.
 */
abstract class Mapper<I, O> {
    abstract fun transform(input: I): O?
    fun trasform(input: List<I>): List<O> {
        val result: MutableList<O> = LinkedList()
        for (inputObj in input) transform(inputObj)?.let(result::add)
        return result
    }

    abstract fun transformBack(input: O): I?
    fun tansformBack(input: List<O>): List<I> {
        val result: MutableList<I> = LinkedList()
        for (inputObj in input) transformBack(inputObj)?.let(result::add)
        return result
    }
}